Add a register_replacement to fix float8 delayed scaling kernel fusion issues

y-sq opened the pull request:
Summary:
We previously tried the `defer_reduction_split_after_fusion` way to fix the fusion issue. 
However, as we agree that the longer-term solution is cooperative reduction + tiled reduction, the defer reduction split approach will also be a shorter-term solution. And we want to keep the shorter-term solution simpler. 

This pr uses the `pattern_matcher` to match the fp8 delayed scaling pattern, simply replace every `max(abs(x))` with `max(abs(x), dim=-1), max()`. It generates the same result as the defer reduction split approach .

Test Plan:
Run float8 training script. Amax and cast are fused in delayed scaling; dynamic scaling is not affected.

The delayed scaling kernel also looks reasonable to me, https://fburl.com/phabricator/iqmlollk

```
TORCH_LOGS="fusion" TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1  buck run mode/opt scripts/shuqiyang/test_inductor:test_float8 --  ~/local/tmp/20241120_test --dtype_filter float8 --scaling_type_input delayed --scaling_type_weight delayed --scaling_type_grad_output delayed  2>&1 | tee ~/test_compile.txt
```

```
TORCH_LOGS="fusion" TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1  buck run mode/opt scripts/shuqiyang/test_inductor:test_float8 --  ~/local/tmp/20241120_test --dtype_filter float8 --scaling_type_input dynamic --scaling_type_weight dynamic --scaling_type_grad_output dynamic  2>&1 | tee ~/test_compile.txt
```

Differential Revision: D67135795




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

y-sq commented on torch/_inductor/fx_passes/misc_patterns.py:
```diff
@@ -69,6 +69,52 @@ def randperm_index_replacement(x, slice_shape):
         scalar_workaround={"slice_shape": 42},
     )
 
+    # Float8 training patterns
```
I can also move the pattern to torchao/float8 folder if we think that's a better place.

eellison replied:
Yea, i think since this is very float8 specific, let's put it there

y-sq commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -1580,7 +1580,7 @@ def gen_pattern(
     return fx_to_pattern(
         search_gm,
-        ignore_types=(int, float, list, torch.device, torch.dtype),
+        ignore_types=(int, float, list, torch.device),
```
If ignore_types contains `torch.dtype`, the fwd (dtype=e4m3) and bwd (dtype=e5m2) patterns will fail at `assert pattern_repr not in _seen_patterns`, https://github.com/pytorch/pytorch/blob/main/torch/_inductor/pattern_matcher.py#L1398

eellison replied:
Don't think we can land this but https://github.com/pytorch/pytorch/pull/139321 should fix

y-sq replied:
thanks for the comment. I'll remove this change and test the pr again with https://github.com/pytorch/pytorch/pull/139321

eellison replied:
landed ! we could allow register_replacement api to specify an override of `ignore_types` - just not make a global change like here. in any case, with the pr, shouldn't need any further change.

y-sq replied:
It works with your pr. thanks!

y-sq commented:
@pytorchbot label "topic: not user facing"

vkuzo commented on torch/_inductor/fx_passes/misc_patterns.py:
```diff
@@ -69,6 +69,52 @@ def randperm_index_replacement(x, slice_shape):
+
+    def amax_with_scaling_pattern(tensor_x_inp, scale_x, IS_BWD):
+        tensor_x = tensor_x_inp.to(torch.float32) * scale_x
+        if IS_BWD:
```
instead of using `IS_BWD`, can we just register replacements for both e4m3 and e5m2 variants in both forward and backward?  Motivation: "e4m3 in fwd, e5m2 in bwd" is only one of the possible recipes.

y-sq replied:
ah yes, the difference in `is_bwd` and `not is_bwd` is actually the different data type, I'll just change the name to `is_e5m2` to make it more clear.

vkuzo commented on torch/_inductor/fx_passes/misc_patterns.py:
```diff
@@ -69,6 +69,52 @@ def randperm_index_replacement(x, slice_shape):
+        for IS_BWD in [True, False]:
+            # torch.bfloat16 has the same pattern as torch.float16, because they both needs `tensor_x_inp.to(torch.float32)` 
+            # Putting both `torch.bfloat16` and `torch.float16`patterns will cause errors in `assert pattern_repr not in _seen_patterns`
+            for dtype in [torch.float32, torch.float16]: 
```
maybe use `bfloat16` instead of `float16` here?

vkuzo commented on torch/_inductor/fx_passes/misc_patterns.py:
```diff
@@ -69,6 +69,52 @@ def randperm_index_replacement(x, slice_shape):
+        amax = torch.max(amax_1)
+        return (tensor_x, amax)
+    
+    # The amax_with_scaling_pattern will also match dynamic scaling cases, we want to avoid that.
```
is this expected?  The pattern for delayed scaling should be different from that of dynamic scaling:

```
# delayed pseudocode
x1, s1 = cast_and_get_new_scale(x0, s0)

# dynamic pseudocode
s1 = get_scale(x0)
x1 = cast(x0, s1)
```

eellison replied:
hmm, yea, it's surprising to me this is occurring. is there any other way to differentiate ? maybe based on uses of the scale, or amax ?

vkuzo commented on torch/_inductor/fx_passes/misc_patterns.py:
```diff
@@ -69,6 +69,52 @@ def randperm_index_replacement(x, slice_shape):
+    # The amax_with_scaling_pattern will also match dynamic scaling cases, we want to avoid that.
+    # `scale_x` of delayed scaling comes from the previous iteration (a `primals`), instead of from a function call.
+    def fp8_delayed_scaling_extra_check(match):
+        return match.kwargs["scale_x"].op != "call_function"
```
not sure if this is going to survive future refactors in the torchao codebase, can we match on the pattern itself instead?

y-sq replied:
Both dynamic and delayed scaling have the following pattern pieces:
```
        tensor_x = tensor_x_inp.to(torch.float32) * scale_x
        tensor_x = tensor_x.clamp(min=-1 * E4M3_MAX_POS, max=E4M3_MAX_POS)
        tensor_x = tensor_x.to(torch.float8_e4m3fn)
        amax = torch.max(torch.abs(tensor_x_inp))
```

The difference is the `scale_x` for delayed scaling comes from the sync_and_update_amax call (which is the input node to the forward/backward graph). But the `scale_x` of dynamic scaling is calculated from ` amax = torch.max(torch.abs(tensor_x_inp))` (a function call).

I am thinking to use another extra_check which checks if the input nodes of `scale_x` contains `tensor_x_inp`, which might work better to differentiate delayed and dynamic scaling. I'll update the pr if it works.

vkuzo replied:
hmm, I think if the logic here ends up being specific to how delayed scaling is implemented in torchao, then maybe that's evidence that torchao is a better place to put this code? Is there is an extension point in inductor where we can register this pass from torchao?

y-sq replied:
Hi @eellison , do you have any suggestions here? maybe I can add a `custom_pattern_replacements` option in config? https://fburl.com/2f4y29hf

y-sq replied:
@vkuzo I updated the logic of `extra_check`. It now checks whether `scale_x` is from `tensor_x_inp`. 
We might update how delayed scaling is implemented in torchao, but it should be always true that "In delayed scaling case, the input of `scale_x` does not contain `tensor_x_inp`", so the check less can be less coupled with torchao implementation.


I also got some performance number. 
As we discussed, we want to make "recompute fp8 weight and activation in backward" the default behavior, I also used these settings to test performance. I temporarily enforced them by using activation checkpointing (to recompute fp8 activation) and the config `force_recompute_fp8_weight_in_bwd`. The local script, D67517255.

```
Shape: 4096x4096x4096
Single linear layer
Before: 0.7000332674418591 ms
After: 0.6707110309278329 ms
Delta: 4.4%

With layer_norm:
Before: 0.753488252631584
After: 0.7374823118279574
Delta: 2.2%

Witi sigmoid:
Before: 0.7153208260869579
After: 0.6845765714285723
Delta: 4.5%
```

y-sq committed:
Add a register_replacement to fix float8 delayed scaling kernel fusion issues (#143464)

Summary:

We previously tried the `defer_reduction_split_after_fusion` way to fix the fusion issue. 
However, as we agree that the longer-term solution is cooperative reduction + tiled reduction, the defer reduction split approach will also be a shorter-term solution. And we want to keep the shorter-term solution simpler. 

This pr uses the `pattern_matcher` to match the fp8 delayed scaling pattern, simply replace every `max(abs(x))` with `max(abs(x), dim=-1), max()`. It generates the same result as the defer reduction split approach .

Test Plan:
Run float8 training script. Amax and cast are fused in delayed scaling; dynamic scaling is not affected.

The delayed scaling kernel also looks reasonable to me, https://fburl.com/phabricator/iqmlollk

```
TORCH_LOGS="fusion" TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1  buck run mode/opt scripts/shuqiyang/test_inductor:test_float8 --  ~/local/tmp/20241120_test --dtype_filter float8 --scaling_type_input delayed --scaling_type_weight delayed --scaling_type_grad_output delayed  2>&1 | tee ~/test_compile.txt
```

```
TORCH_LOGS="fusion" TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1  buck run mode/opt scripts/shuqiyang/test_inductor:test_float8 --  ~/local/tmp/20241120_test --dtype_filter float8 --scaling_type_input dynamic --scaling_type_weight dynamic --scaling_type_grad_output dynamic  2>&1 | tee ~/test_compile.txt
```

Also tested with preceding ops,
```
TORCH_LOGS="fusion" TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1  buck run mode/opt scripts/shuqiyang/test_inductor:test_float8 --  ~/local/tmp/20241120_test --dtype_filter float8 --scaling_type_input delayed --scaling_type_weight delayed --scaling_type_grad_output delayed  --model_type ln_linear 2>&1 | tee ~/test_compile.txt
```

A simple performance test: D67517255

Differential Revision: D67135795

eellison commented on torch/_inductor/fx_passes/misc_patterns.py:
```diff
@@ -69,6 +71,71 @@ def randperm_index_replacement(x, slice_shape):
+
+    if torch.cuda.is_available():
+        for IS_E5M2 in [True, False]:
+            # torch.float16 has the same pattern as torch.bfloat16, because they both needs `tensor_x_inp.to(torch.float32)`
+            # It will cause errors in `assert pattern_repr not in _seen_patterns`
```
this comment no longer relevant ?

eellison reviewed (commented):
Looks good as a workaround - I think it might be better to put in float8 but will defer to @vkuzo
